[java] new plugin: get information about installed JARs

msrb opened the pull request:
This plugin collects information about JAR files installed in the
standard locations. The information can be later used when debugging
problems with missing/different than expected classes on classpath.

Signed-off-by: Michal Srb <msrb@redhat.com>

msrb commented:
Oops, pep8 :) I will fix it.

msrb commented:
Actually I think I can improve this PR a bit. Temporarily marking as WIP.

bmr-cymru commented:
No problem - feel free to add additional commits to the branch, or to force-push an updated branch with revised commits.

I had a quick look through and mostly looks sane - the only "wrong" thing I saw was that your subject tag has `[java] new plugin` - conventionally in sos that means the commit adds a new plugin named `java`, whereas here the plugin is called `jars`. I'd be happy either just fixing this up, or if you want, merging it into the existing Java plugin (which is very small/simple currently - just collects a few files and commands).

The only other thing worth pointing out is that we've been debating switching to more structured data in the reports for some time - I see you're using JSON for all of the output of this plugin - that's not a problem, but if we did go down that route in the future it may mean moving things around, or changing the plugin behavior in future: today most plugins log plain text and there's been discussion about adding a structured format and then allowing the "legacy" formats to be optionally disabled.

Anyway: no final decisions have been made yes (but I'd be very surprised if we did implement this, and chose to use anything other than JSON for the serialisation format).

msrb commented:
Hi Bryn,

Thanks for your comment and sorry for the delay. Here's what I've changed:
- added option to specify additional locations where to look for JARs
- fixed subject tag (`[java]` -> `[jars]`)

With regard to JSON, I would prefer to keep it this way if it's not a problem. Of course, I wouldn't be opposed to any changes in the future, if required.

msrb commented:
Hi Bryn, sorry for being impatient here :) Is there anything I could do to help to get this PR merged? Thanks.

(cc @krishnapaparaju - this is the Java plugin I was talking about)

bmr-cymru commented:
@msrb, been out on vacation - I expect to do a round of merges either end of this week, or the beginning of next - this PR should be an uncomplicated merge.

bmr-cymru commented:
Uggh, sorry - got tied up with work on other projects. Catching up on this PR now.

bmr-cymru commented:
My only remaining concern with this plugin is whether it has the potential to become expensive on some systems: on my laptop (reasonably fast SSD, plenty of RAM), and with ~1400 JARs present on the system, it adds around 5s to a run. That's not too bad but it would be good to have some more data on this before `3.4` is out.

bmr-cymru commented:
Closed in commit 3dfcdac0e3e0c42337eb846534b323a739bfc068, thanks!

If you have any time to test and gather more data from systems with very many JARs, or with very large `/opt` trees, that would be very useful and helps ensure we don't have any surprises when this is released to distro users.

bmr-cymru commented:
We have reports of performance problems with this plugin and large collections of data under `/var/lib`; scanning of this entire location must be disabled by default. If there are known paths under `lib` that can be safely trawled then we an whitelist those.

msrb commented:
Ouch, I am really sorry about that. I've never had any performance problem myself, but it's true that `/var/lib` could be pretty populated on some systems.

It's perfectly fine to remove `/var/lib` from the list of default locations. I'd maybe go even further and remove `/opt` and `/usr/local` as well - just to play it safe. The plugin can take additional option `jars.append_locations=...` where people can specify more locations, when necessary.

I will open a PR shorty.

bmr-cymru commented:
The current plan is to only scan the paths in `/usr`, and to have a white-listed set of additional paths that the user may optionally enable (in addition to specifying particular locations via the existing plugin option).
